### PR TITLE
ENH: PyDMSlider Parameters Menu

### LINKED ISSUE #826: Adjust slider step size at run time
The EDM slider widget has a feature where you can right-click it, and a little pop-up will appear that lets you adjust the slider step size, and also just type in a value to go to.  This is pretty useful - it is pretty difficult to set a slider step size that is always the right choice.  If your accelerator's beam energy changes to something much lower, those gentle corrector magnet steps suddenly seem a lot less gentle...

### PR DESCRIPTION
**Pull request by YektaY**
## Description
Adds a menu which can be accessed by right clicking on the slider. 

The menu can modify a set of parameters: 
            1)	value
            2)	new step size (float or channel) and scale factor for step size
            3)	precision or if precision is defined from a channel
            4)	format of numbers on slider (min, max, value) - float or exp

For the step size a int/float or a channel can be given.

One quirk of the step size, is that if the given step size is not commensurate to the given range (min to max) then the last step will be smaller then the previous steps since I have it set to the max value. Currently this fact is not noted to the user and simple handled silently in the background. 

Another design decision which could be argued, currently if bad user input is given, the program will silently reject the input (unless someone is looking at the logs).   

Also not sure if folks like the 1e0 format instead of say x10^0 
  
## Motivation and Context
Addresses an open request. #826 

## How Has This Been Tested?
Tested with the Slider example in PyDM

## Where Has This Been Documented?
No new documentation has been made at the moment

## Screenshots:
Screen Shot 2022-08-25 at 4 50 25 PM

### CODE REVIEW

File: pydm/widgets/slider.py [hunk 1]
```diff
@@ -73,14 +80,157 @@ def __init__(self, parent=None, init_channel=None):
...
+
+    def apply_step_size_menu_changes(self):
+        """
+        Method which attempts to set the user imputed data from the slider parameters menu.
+        """
+        try:
+            val = float(self.input[1].text())
+            val = val*float(self.input[2].currentText())
+            if val > 0:
+                self.step_size = val
```
Comment (jbellister-slac) [inline pydm/widgets/slider.py:201]: A more descriptive name than val would help here (and perhaps later in this method as well). It's hard to tell what is actually being multiplied here without running the code and opening the menu.

**Review by jbellister-slac — COMMENTED** [review]
This is looking very nice, will be a big improvement to this widget!

I'm trying some functionality using the example slider.ui from PyDM examples, and seeing this odd behavior when just trying to set the step size and it actually changes the value of the PV as well (a camonitor show it is being written to):

slider

If you can't reproduce this on your end let me know and we can take a look in person.

I think it would also be worth adding a test case or two to the `test_slider` file that covers some of the more important logic around the new code you wrote

File: pydm/widgets/slider.py [hunk 2]
```diff
@@ -73,14 +80,157 @@ def __init__(self, parent=None, init_channel=None):
         self.setup_widgets_for_orientation(self._orientation)
         self.reset_slider_limits()
 
+        self.widget = None
+        self.input = None
+        self.label = None
+        self.button = None
```
Comment (jbellister-slac) [inline pydm/widgets/slider.py:86]: Would also help to give more descriptive names to these as well so it is easier to know what their purpose is.

File: pydm/widgets/slider.py [hunk 3]
```diff
@@ -73,14 +80,157 @@ def __init__(self, parent=None, init_channel=None):
...
+        """
+        Method that builds a menu to modify a set of Slider Parameters:
+            1)	value
+            2)	new step size (float or channel) and scale factor for step size
+            3)	precision or if precision is defined from a channel
+            4)	format of numbers on slider (min, max, value) - float or exp
+        """
+        self.widget = QWidget()
+        self.widget.show()
+
```
[the remark this answers is not in the document]
Comment (YektaY) [reply pydm/widgets/slider.py:122]: great idea. I will implement this suggestion. Thank you.

**Review by jbellister-slac — APPROVED** [review]
Thanks for the updates, works as expected and looks good to me!

**Pull request merged**